### ISSUE TITLE
build dependencies not working on osx

### ISSUE THREAD

**Issue by drjrkuhn** [JuliaGL/GLFW.jl#94]
The build process for GLFW is not working on the new macOS Sierra. I have spent some time tracking down the reason, and I believe I have found it:

The [GLFW.jl/deps/build.jl](https://github.com/JuliaGL/GLFW.jl/blob/master/deps/build.jl) file runs and downloads the prebuilt glfw3 homebrew package, but the package cannot find the correct `opengl32` dynamic library when running. With every new OS release, the default framework path changes. For example, the new framework path for opengl on sierra has moved to `/Applications/Xcode.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX10.12.sdk/System/Library/Frameworks/OpenGL.framework` (*note the 10.12*).

The problem is that the prebuilt bottles referenced in `deps/build.jl` link to their own specific framework folder. This mechanism falls behind with each new OS release. The regular command line homebrew manager knows this and will build glfw3 from source rather than download a bottle. The mechanism used in GLFW.jl bypasses homebrew and cannot handle the change properly. Since all of the JuliaGL packages at their root appear to rely on GLFW for the binary dependencies to opengl, fixing it here should fix it everywhere else.

I propose that the `deps/build.jl` process be reworked to use BinDeps.jl and Homebrew.jl rather than relying on its own mechanism of downloading prebuilt bottles to GLFW/deps/builds. Perhaps something more along the lines of the deps found in [Cairo.jl/deps/build.jl](https://github.com/JuliaGraphics/Cairo.jl/blob/master/deps/build.jl). Unfortunately, that likely means moving the libglfw3 binary files out of their GLFW/deps/builds subfolder and into the regular BinDeps.jl and Homebrew.jl target paths.

**Comment by SimonDanisch** [comment]
I'm not sure about what you write about the built, but to me it sounds as if you run into: https://github.com/JuliaGL/GLVisualize.jl/issues/116

**Comment by drjrkuhn** [comment]
I believe the issue in GLVisualize.jl is really an issue with GLFW.jl, since I get the same error of opening "opengl32" when running either Pkg.test("GLFW") or Pkg.test("GLVisualize") or Pkg.test("GLWindow") or even Pkg.test("ModernGL").

- Pkg.test("GLVisualize") reports
> ERROR: LoadError: error compiling #create_glcontext#42: error compiling glDebugMessageCallbackARB: error compiling getprocaddress_e: error compiling getprocaddress: could not load library "opengl32"
dlopen(opengl32.dylib, 1): image not found
 in (::GLWindow.#kw##create_glcontext)(::Array{Any,1}, ::GLWindow.#create_glcontext, ::String) at ./<missing>:0 
...

- Pkg.test("ModernGL") reports
> ERROR: LoadError: error compiling anonymous: error compiling createcontextinfo: error compiling glGetString: error compiling getprocaddress_e: error compiling getprocaddress: could not load library "opengl32"
...


- Pkg.test("GLWindow") reports an error in create_glcontext, which calls GLFW
> ERROR: LoadError: error compiling #create_glcontext#42: error compiling glDebugMessageCallbackARB: error compiling getprocaddress_e: error compiling getprocaddress: could not load library "opengl32"
dlopen(opengl32.dylib, 1): image not found
 in (::GLWindow.#kw##create_glcontext)(::Array{Any,1}, ::GLWindow.#create_glcontext, ::String) at ./<missing>:0
...

- Pkg.test("GLFW") reports
> ERROR: LoadError: LoadError: error compiling glGenVertexArrays: error compiling getprocaddress_e: error compiling getprocaddress: could not load library "opengl32"
dlopen(opengl32.dylib, 1): image not found
 in include_from_node1(::String) at ./loading.jl:488
 ...

GLVisualize calls ModernGL.getprocaddress, which calls GLFW for the context, which triggers loading of opengl32.dylib. Similarly, GLWindow also calls GLFW for a context which tries to load opengl32.dylib.

To get GLFW to work, the homebrew package needs to be rebuilt if the proper bottle for v"10.12" does not exist during the build step in GLFW.jl/deps/build.jl  `if is_apple()` section.

**Comment by drjrkuhn** [comment]
Actually, I believe this is a separate issue from the one in https://github.com/JuliaGL/GLVisualize.jl/issues/116.

**Comment by jayschwa** [comment]
The quick fix is to add an extra line that downloads the bottle for Sierra. I presume one is available in Homebrew's appveyor.

Agreed that it would be nice to leverage Homebrew instead. I made an attempt a couple years ago, but found it was harder than I expected for reasons I no longer remember.

**Comment by drjrkuhn** [comment]
There is not a bottle for sierra yet. I have been checking for the last few days.

**Comment by jayschwa** [comment]
Weird. I expected that would happen auto-magically. Maybe the library's formula doesn't build cleanly on it.

**Comment by drjrkuhn** [comment]
from [Homebrew/brew/docs/Bottles.md](https://github.com/Homebrew/brew/blob/bbed7246bc5c5b7acb8c1d427d10b43e090dfd39/docs/Bottles.md#bottle-creation)

Bottle creation is not done automatically. The glfw3 package maintainer has to update the formula `bottle do` block and then submit a pull request before a new bottle is built by the bot.

**Comment by SimonDanisch** [comment]
So have you tried checking out ModernGL? Since GLVisualize, GLWindow, ModernGL and GLFW rely on the same feature in ModernGL for the tests, they should fail with the same error.... As far as I know, there is no `opengl32` dylib in OSX, so the only way it could get in here is over that ModernGL bug....

**Comment by SimonDanisch** [comment]
Please reopen if `Pkg.checkout("ModernGL")` doesn't fix this.

**Comment by drjrkuhn** [comment]
Yes. Thanks!

After dumping all of my julia libraries and doing a `Pkg.update()` to repopulate the package folders, I tried a `Pkg.checkout("ModernGL")` followed by `Pkg.test("GLFW")`. The tests ran OK after that. GLVisualize also tested OK. Thanks for your help!

Before I tried that, I made a stab at using rewriting GLFW/deps/build.jl to use BinDeps and Homebrew to download libglfw3 in a more "regular" way. I based it on the [GSL.jl example](https://github.com/jiahao/GSL.jl/blob/master/deps/build.jl) recommended by BinDeps. I forked GLFW.jl  to [github.com/drjrkuhn/GLFW.jl](https://github.com/drjrkuhn/GLFW.jl) to try it out. The OSX version builds and tests find. The binaries are placed in the Homebrew package folder instead of the GLFW package folder. 

Unfortunately, the fork is not yet ready for prime time. I tried the same fork of GLFW on an Ubuntu machine and ran into a bug in BinDeps where it tries to install packages in `sudo` mode.
